src/libfsm/intersect.c

```diff
@@ -65,3 +72,317 @@ fsm_intersect(struct fsm *a, struct fsm *b)
 	return NULL;
 }
 
+struct pair {
+	struct fsm_state *a;
+	struct fsm_state *b;
+	struct fsm_state *comb;
+};
+
+static int cmp_pair(const void *a, const void *b)
+{
+	const struct pair *pa = a, *pb = b;
+	ptrdiff_t delta;
+	delta = pa->a - pb->a;
+	if (delta < 0) {
+		return -1;
+	}
+	
+	if (delta > 0) {
+		return 1;
+	}
+
+	delta = pa->b - pb->b;
+	if (delta < 0) {
+		return -1;
+	}
+
+	if (delta > 0) {
+		return 1;
+	}
+
+	return 0;
+}
+
+struct pair_pool {
+	struct pair_pool *next;
+	struct pair items[PAIR_POOL_SIZE];
+};
```

Comment:
It's not really a cache.  Nor does it serve the same purpose as a set.

It's just for allocation and cleanup.  the bywalk algorithm needs to allocate a bunch of (A,B,combined) triples and then free them when finished (or in case of an error).  I wanted to make cleanup easier, so instead of hunting down all the triples, I allocate them from a  pool and then free the entire pool at the end.

Reply:
It should have the small added benefit of reducing malloc() fragmentation, but it's mainly to make cleanup easier.